stats/view/aggregation_data.go

```diff
@@ -139,26 +159,61 @@ func (a *DistributionData) addSample(f float64) {
 	a.SumOfSquaredDev = a.SumOfSquaredDev + (f-oldMean)*(f-a.Mean)
 }
 
-func (a *DistributionData) incrementBucketCount(f float64) {
-	if len(a.bounds) == 0 {
-		a.CountPerBucket[0]++
+func (a *DistributionData) addToBucket(val float64, ex *exemplar.Exemplar) {
+	var (
+		count  *int64
+		bucket *Bucket
+	)
+
+	for i, b := range a.bounds {
+		if val < b {
+			count = &a.CountPerBucket[i]
+			bucket = &a.Buckets[i]
+			break
+		}
+	}
+
+	if count == nil {
+		count = &a.CountPerBucket[len(a.bounds)]
+		bucket = &a.Buckets[len(a.bounds)]
+	}
+
+	bucket.Count++
+	*count = bucket.Count
+	bucket.maybeRetainExemplar(val, ex)
+}
+
+func (b *Bucket) maybeRetainExemplar(val float64, e *exemplar.Exemplar) {
+	if b.Exemplar == nil {
+		b.Exemplar = &ExemplarData{
+			Value:       val,
+			Timestamp:   e.Timestamp,
+			Attachments: e.Attachments,
+		}
 		return
 	}
 
-	for i, b := range a.bounds {
-		if f < b {
-			a.CountPerBucket[i]++
-			return
+	// Heuristic to pick the "better" exemplar: first keep the one with a
+	// sampled trace attachment, if neither have a trace attachment, pick the
+	// one with more attachments.
+	_, haveTraceID := e.Attachments["TraceID"]
+	if haveTraceID || len(e.Attachments) >= len(b.Exemplar.Attachments) {
+		b.Exemplar = &ExemplarData{
+			Value:       val,
+			Timestamp:   e.Timestamp,
+			Attachments: e.Attachments,
 		}
 	}
-	a.CountPerBucket[len(a.bounds)]++
 }
 
 func (a *DistributionData) clone() AggregationData {
+	c := *a
 	counts := make([]int64, len(a.CountPerBucket))
 	copy(counts, a.CountPerBucket)
-	c := *a
 	c.CountPerBucket = counts
+	buckets := make([]Bucket, len(a.Buckets))
+	copy(buckets, a.Buckets)
```

Comment:
This only changes the slices but the underlying ExemplarData per bucket is the same
since the ExemplarData pointer is unchanged so this is a shallow copy. Copying CountsPerBucket copies int64s which are just values.

Reply:
Yeah but we don't expect anyone to mutate the map at this stage (it's only mutated when the exemplary is first created) so I think shallow copying in this case is fine.

Reply:
I created a type for attachments and added this as a note on the type godoc.